go-fuzz-build/pkgs.go

```diff
@@ -0,0 +1,32 @@
+// Copyright 2015 go-fuzz project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/tools/go/packages"
+)
+
+// packageDir returns local directory with package source files.
+func packageDir(p *packages.Package) string {
+	// Go-package contains at least one go-file, so GoFiles is not empty without fail.
+	dir := filepath.Dir(p.GoFiles[0])
+	return dir
+}
+
+// isNotPackage checks if dir contains go source files.
```

Comment:
Negatives are harder to read at call sites. I'd prefer `isPackage`, and using `!isPackage` at call sites if needed. This comment should read: `// isPackage reports whether dir contains Go source files.`. However, that's not really the right condition, because whether a Go source file is relevant depends on whether it is excluded by current build tags, etc. It'd be better to rely on go/packages to tell us whether a directory is a package or not. If it is a package that we care about, then go/packages will return a Package for it. So perhaps a better implementation is to extract filepath.Dir for every Package returned by go/packages and then check whether dir is present in that set.

The reason I'm not sure whether that's a better implementation is that I still don't understand *why* this PR is necessary. The description claims that it is necessary, but it doesn't explain:

* why we should expect non-package directories to be important (what do they contain)
* where we should look for them (all subdirs of package dirs? recursively in all subdirs of package dirs?) and why we should look for them there
* how to tell when we should be looking for non-package directories
* why this won't cause us to accidentally copy large unnecessary unrelated data directories that might be present in other code bases
* why this needs to be fixed in go-fuzz-build instead of in the code base that encountered the issue

Before writing any more code, please answer this set of questions, because it will guide what the code should do. Thank you.

Reply:
> why we should expect non-package directories to be important (what do they contain) ?

As it was shown by [example](https://github.com/ethereum/go-ethereum/tree/v1.9.22/crypto/secp256k1/libsecp256k1/incude), such non-package dirs could contain C-code files. And go-code includes they by CGO.

Reply:
> where we should look for them (all subdirs of package dirs? recursively in all subdirs of package dirs?) and why we should look for them there?

There is no 100% right answer - they could be anywhere. I use the assumption that all non-go code the go-repo needs is in git-repo. So we should copy all subdirs tree of the repo. But it leads us to the next question:

> why this won't cause us to accidentally copy large unnecessary unrelated data directories that might be present in other code bases?

Of cause, this could cause to. But it is better than build fail. I try to minimize data copying by crawling only by non-package subdirs. It is next level assumption and compromise.

Reply:
> how to tell when we should be looking for non-package directories?

We cannot anticipate all cases when non-package dirs are neccesary for build, so we must always copy them.

Reply:
> why this needs to be fixed in go-fuzz-build instead of in the code base that encountered the issue?

The [code base](https://github.com/ethereum/go-ethereum/blob/v1.9.22/crypto/secp256k1/secp256.go#L26) that encountered the issue is in his own right. Go-code can include C-code. CGO is a property of the go.

Reply:
Did my answers explain the change reason?